### PR TITLE
feat(TDI-41953): snowflake key pair auth

### PR DESCRIPTION
**Pull request by mbasiuk-talend**
**What is the current behavior?** (You can also link to an open issue here)
https://jira.talendforge.org/browse/TDI-41953
Only basic Authentication type is used

**What is the new behavior?**
Extend authentication types by using Key Pair auth.

**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?
- [ ] The code coverage on new code >75%
- [ ] The new code does not introduce new technical issues (sonar / eslint)

**What kind of change does this PR introduce?**

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other Information**:

### CODE REVIEW

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/157/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/157/)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/158/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/158/)

File: components/components-snowflake/components-snowflake-definition/src/main/java/org/talend/components/snowflake/SnowflakeConnectionProperties.java [hunk 1]
```diff
@@ -221,7 +250,7 @@ public SnowflakeConnectionProperties getReferencedConnectionProperties() {
         return null;
     }
 
-    public Properties getJdbcProperties() {
+    public Properties getJdbcProperties() throws Exception {
```
Comment (undx) [inline components/components-snowflake/components-snowflake-definition/src/main/java/org/talend/components/snowflake/SnowflakeConnectionProperties.java]: Can throws an exception now?

File: components/components-snowflake/components-snowflake-definition/src/main/java/org/talend/components/snowflake/SnowflakeConnectionProperties.java [hunk 2]
```diff
@@ -231,9 +260,14 @@ public Properties getJdbcProperties() {
         if (user != null) {
             properties.put("user", user);
         }
-
-        if (password != null) {
-            properties.put("password", password);
+        if (AuthenticationType.BASIC == authenticationType.getValue()) {
+            if (password != null) {
```
Comment (undx) [inline components/components-snowflake/components-snowflake-definition/src/main/java/org/talend/components/snowflake/SnowflakeConnectionProperties.java]: tests can be combined

File: components/components-snowflake/components-snowflake-definition/src/main/java/org/talend/components/snowflake/SnowflakeConnectionProperties.java (hunk 1 above)
Comment (undx) [reply components/components-snowflake/components-snowflake-definition/src/main/java/org/talend/components/snowflake/SnowflakeConnectionProperties.java]: hum, coming from the getPrivateKey, that's it? (Exception <> IOException)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/159/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/159/)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/160/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/160/)

File: components/components-snowflake/components-snowflake-definition/src/main/java/org/talend/components/snowflake/SnowflakeConnectionProperties.java (hunk 2 above)
Comment (mbasiuk-talend) [reply components/components-snowflake/components-snowflake-definition/src/main/java/org/talend/components/snowflake/SnowflakeConnectionProperties.java]: what do you mean *tests*- condition?

File: components/components-snowflake/components-snowflake-definition/src/main/java/org/talend/components/snowflake/SnowflakeConnectionProperties.java (hunk 1 above)
Comment (mbasiuk-talend) [reply components/components-snowflake/components-snowflake-definition/src/main/java/org/talend/components/snowflake/SnowflakeConnectionProperties.java]: Agree to reduce type.

File: components/components-snowflake/components-snowflake-definition/src/main/java/org/talend/components/snowflake/SnowflakeConnectionProperties.java (hunk 2 above)
Comment (undx) [reply components/components-snowflake/components-snowflake-definition/src/main/java/org/talend/components/snowflake/SnowflakeConnectionProperties.java]: changes did the trick

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/162/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/162/)

File: components/components-snowflake/components-snowflake-integration/src/test/java/org/talend/components/snowflake/test/SnowflakeWritersTestIT.java [hunk 3]
```diff
@@ -254,6 +253,14 @@ public void testOutputFeedback() throws Throwable {
         assertEquals("1", rej.get(6)); // line
         assertEquals("100035", rej.get(8)); // code
 
+        // There could be changes in Snowflake to restrict NULL values in primary key column, so we have to conform this
+        rej = it.next();
+        assertEquals("1", rej.get(1)); // row
+        assertEquals("1", rej.get(3)); // character
+        assertThat((String) rej.get(4), containsString("NULL result in a non-nullable column"));
```
Comment (undx) [inline components/components-snowflake/components-snowflake-integration/src/test/java/org/talend/components/snowflake/test/SnowflakeWritersTestIT.java]: no potential i18n errors here? (just asking ;-) )

Comment (mbasiuk-talend) [reply components/components-snowflake/components-snowflake-integration/src/test/java/org/talend/components/snowflake/test/SnowflakeWritersTestIT.java]: It's rebased version. Sorry, I don't have enough info here(

File: components/components-snowflake/components-snowflake-definition/src/main/java/org/talend/components/snowflake/SnowflakeConnectionProperties.java [hunk 4]
```diff
@@ -264,7 +317,7 @@ public String getConnectionUrl() {
 
         StringBuilder url = new StringBuilder().append("jdbc:snowflake://").append(account);
         url.append(".snowflakecomputing.com")
-                .append("/?");
+        .append("/?");
```
[the remark this answers is not in the document]
Comment (mbasiuk-talend) [reply components/components-snowflake/components-snowflake-definition/src/main/java/org/talend/components/snowflake/SnowflakeConnectionProperties.java]: maybe formatter issue. Need to double check style.

**Pull request merged**